src/witan/app/data.clj

```diff
@@ -117,10 +124,31 @@
     (c/exec (update-version-number-name name version))
     (first (c/exec (find-data-by-data-id data-id)))))
 
-(defresource data [{:keys [category]}]
+(defresource search [{:keys [category]}]
   util/json-resource
   :allow-methods #{:get}
   :handle-ok (fn [ctx]
                (s/validate [ws/DataItem] (map #(->Data % true) (get-data-by-category
                                                                 category
                                                                 (util/get-user-id ctx))))))
+
+(defresource data [{:keys [category name file public user-id]}]
+  :allowed-methods #{:post}
+  :available-media-types ["application/json"]
+  :processable? (fn [ctx]
+                  (log/info public)
+                  (cond
+                    (not (validation/csv-extension? (:filename file))) [false {:error "this doesn't seem to be a csv file"}]
+                    :default (validation/validate-content category (:tempfile file))))
+  :handle-unprocessable-entity (fn [ctx] {:error (:error ctx)})
+  :handle-created (fn [ctx]
+                    (let [s3-key (java.util.UUID/randomUUID)
+                          post-params (util/get-post-params ctx)]
+                      (try (s3/upload s3-key (:tempfile file))
```

Comment:
I think it'd be more appropriate (in terms of liberator semantics) to do the upload in `post!` rather than `handle-created`